{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2c2cd102",
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract expression values for each transcript in chess for each tissue\n",
    "\n",
    "# format of the input file\n",
    "# ALL_03716214\tSRR1466260:0.652967\n",
    "# ALL_02055884\tSRR1445170:0.560959\tSRR661373:1.153505\tSRR818850:0.479766\n",
    "# ALL_10044090\tSRR1069734:2.544744\n",
    "# ALL_18714109\tSRR1101373:0.678579\n",
    "# ALL_17680799\tSRR1081116:1.300604\tSRR1356643:1.430289\n",
    "# ALL_25719799\tSRR613963:1.111076\n",
    "# ALL_13221907\tSRR1343053:3.592212\n",
    "# ALL_22293686\tSRR820733:0.413197\n",
    "# ALL_08997145\tSRR1079046:1.633669\tSRR1090070:2.432191\tSRR1351390:2.003746\tSRR1365013:1.225308\tSRR1442592:0.674933\tSRR1456693:0.899244\tSRR659637:0.5877\n",
    "# ALL_04709978\tSRR1436858:2.427112\tSRR615443:2.889572\tSRR819771:0.835807\n",
    "\n",
    "# we can extract chess id to all id mapping from the chess gtf file\n",
    "\n",
    "# then we also get the list of sample ids to bodysites and tissues as follows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0754656f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import time\n",
    "import glob\n",
    "import shutil\n",
    "import argparse\n",
    "import subprocess\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ce36ac96",
   "metadata": {},
   "outputs": [],
   "source": [
    "non_zero_tpms_fname = \"/ccb/salz8-1/avaraby/chess3.1/data/ALL.expression.nonzero.tpms\"\n",
    "chess_gtf_fname = \"/home/avaraby1/genomes/human/hg38/annotations/chess3.1.3.GRCh38.gtf\"\n",
    "info_fname = \"/ccb/salz8-1/avaraby/GTEx_metadata/GTEx_salmples_info.tab\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a9ca6e2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>MBases</th>\n",
       "      <th>MBytes</th>\n",
       "      <th>Run</th>\n",
       "      <th>body_site</th>\n",
       "      <th>histological_type</th>\n",
       "      <th>is_tumor</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PAIRED</td>\n",
       "      <td>5876</td>\n",
       "      <td>2417</td>\n",
       "      <td>SRR1098761</td>\n",
       "      <td>Cells - Cultured fibroblasts</td>\n",
       "      <td>Skin</td>\n",
       "      <td>No</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PAIRED</td>\n",
       "      <td>4833</td>\n",
       "      <td>2206</td>\n",
       "      <td>SRR1098785</td>\n",
       "      <td>Lung</td>\n",
       "      <td>Lung</td>\n",
       "      <td>No</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PAIRED</td>\n",
       "      <td>5072</td>\n",
       "      <td>2326</td>\n",
       "      <td>SRR1098807</td>\n",
       "      <td>Ovary</td>\n",
       "      <td>Ovary</td>\n",
       "      <td>No</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PAIRED</td>\n",
       "      <td>4449</td>\n",
       "      <td>1864</td>\n",
       "      <td>SRR1098831</td>\n",
       "      <td>Esophagus - Muscularis</td>\n",
       "      <td>Esophagus</td>\n",
       "      <td>No</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PAIRED</td>\n",
       "      <td>5737</td>\n",
       "      <td>2709</td>\n",
       "      <td>SRR1098855</td>\n",
       "      <td>Stomach</td>\n",
       "      <td>Stomach</td>\n",
       "      <td>No</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24601</th>\n",
       "      <td>PAIRED</td>\n",
       "      <td>5739</td>\n",
       "      <td>2444</td>\n",
       "      <td>SRR1094168</td>\n",
       "      <td>Cells - Transformed fibroblasts</td>\n",
       "      <td>Skin</td>\n",
       "      <td>No</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24602</th>\n",
       "      <td>PAIRED</td>\n",
       "      <td>4944</td>\n",
       "      <td>2150</td>\n",
       "      <td>SRR1097076</td>\n",
       "      <td>Esophagus - Muscularis</td>\n",
       "      <td>Esophagus</td>\n",
       "      <td>No</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24603</th>\n",
       "      <td>PAIRED</td>\n",
       "      <td>4403</td>\n",
       "      <td>1887</td>\n",
       "      <td>SRR1097905</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>Prostate</td>\n",
       "      <td>No</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24604</th>\n",
       "      <td>PAIRED</td>\n",
       "      <td>5394</td>\n",
       "      <td>2311</td>\n",
       "      <td>SRR1100564</td>\n",
       "      <td>Artery - Aorta</td>\n",
       "      <td>Blood Vessel</td>\n",
       "      <td>No</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24605</th>\n",
       "      <td>PAIRED</td>\n",
       "      <td>4883</td>\n",
       "      <td>2125</td>\n",
       "      <td>SRR1105417</td>\n",
       "      <td>Esophagus - Mucosa</td>\n",
       "      <td>Esophagus</td>\n",
       "      <td>No</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24606 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      LibraryLayout  MBases  MBytes         Run  \\\n",
       "0            PAIRED    5876    2417  SRR1098761   \n",
       "1            PAIRED    4833    2206  SRR1098785   \n",
       "2            PAIRED    5072    2326  SRR1098807   \n",
       "3            PAIRED    4449    1864  SRR1098831   \n",
       "4            PAIRED    5737    2709  SRR1098855   \n",
       "...             ...     ...     ...         ...   \n",
       "24601        PAIRED    5739    2444  SRR1094168   \n",
       "24602        PAIRED    4944    2150  SRR1097076   \n",
       "24603        PAIRED    4403    1887  SRR1097905   \n",
       "24604        PAIRED    5394    2311  SRR1100564   \n",
       "24605        PAIRED    4883    2125  SRR1105417   \n",
       "\n",
       "                             body_site histological_type is_tumor     sex  \n",
       "0         Cells - Cultured fibroblasts              Skin       No    male  \n",
       "1                                 Lung              Lung       No    male  \n",
       "2                                Ovary             Ovary       No  female  \n",
       "3               Esophagus - Muscularis         Esophagus       No  female  \n",
       "4                              Stomach           Stomach       No  female  \n",
       "...                                ...               ...      ...     ...  \n",
       "24601  Cells - Transformed fibroblasts              Skin       No    male  \n",
       "24602           Esophagus - Muscularis         Esophagus       No    male  \n",
       "24603                         Prostate          Prostate       No    male  \n",
       "24604                   Artery - Aorta      Blood Vessel       No    male  \n",
       "24605               Esophagus - Mucosa         Esophagus       No    male  \n",
       "\n",
       "[24606 rows x 8 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load info\n",
    "info = pd.read_csv(info_fname,sep=\"\\t\")\n",
    "info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "40301931",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of bodysites: 55\n",
      "Cells_Cultured.fibroblasts 666\n",
      "Lung 891\n",
      "Ovary 254\n",
      "Esophagus_Muscularis 699\n",
      "Stomach 487\n",
      "Muscle_Skeletal 1105\n",
      "Skin_Sun.Exposed.Lower.leg 928\n",
      "Pancreas 469\n",
      "Cells_EBV-transformed.lymphocytes 285\n",
      "Esophagus_Mucosa 788\n",
      "Brain_Putamen.basal.ganglia 243\n",
      "Adipose_Subcutaneous 871\n",
      "Brain_Nucleus.accumbens.basal.ganglia 290\n",
      "Artery_Tibial 827\n",
      "Whole.Blood 2797\n",
      "Colon_Transverse 500\n",
      "Adrenal.Gland 368\n",
      "Prostate 282\n",
      "Esophagus_Gastroesophageal.Junction 440\n",
      "Brain_Frontal.Cortex.BA9 267\n",
      "Thyroid 858\n",
      "Adipose_Visceral.Omentum 598\n",
      "Artery_Aorta 568\n",
      "Uterus 216\n",
      "Small.Intestine_Terminal.Ileum 249\n",
      "Bladder 24\n",
      "Breast_Mammary.Tissue 530\n",
      "Testis 493\n",
      "Minor.Salivary.Gland 175\n",
      "Skin_Not.Sun.Exposed.Suprapubic 683\n",
      "Brain_Cerebellar.Hemisphere 276\n",
      "Liver 332\n",
      "Vagina 227\n",
      "Brain_Anterior.cingulate.cortex.BA24 236\n",
      "Pituitary 324\n",
      "Fallopian.Tube 14\n",
      "Spleen 290\n",
      "Colon_Sigmoid 429\n",
      "Cells_Leukemia.cell.line.CML 248\n",
      "Artery_Coronary 325\n",
      "Heart_Atrial.Appendage 531\n",
      "Kidney_Cortex 88\n",
      "Brain_Substantia.nigra 171\n",
      "Brain_Cortex 311\n",
      "Brain_Amygdala 193\n",
      "Brain_Caudate.basal.ganglia 316\n",
      "Brain_Hypothalamus 244\n",
      "Brain_Hippocampus 245\n",
      "Brain_Cerebellum 344\n",
      "Nerve_Tibial 782\n",
      "Brain_Spinal.cord.cervical.c-1 178\n",
      "Heart_Left.Ventricle 628\n",
      "Cervix_Endocervix 10\n",
      "Cervix_Ectocervix 12\n",
      "Cells_Transformed.fibroblasts 1\n"
     ]
    }
   ],
   "source": [
    "# group samples by bodysite instead of tissue\n",
    "\n",
    "bodysite_to_run = dict()\n",
    "run_to_bodysite = dict()\n",
    "tissue_to_bodysite = dict()\n",
    "bodysite_to_tissue = dict()\n",
    "\n",
    "for idx,row in info.iterrows():\n",
    "    bs = row[\"body_site\"].replace(\" - \",\"_\").replace(\" \",\".\").replace(\"(\",\"\").replace(\")\",\"\")\n",
    "    hist_type = row[\"histological_type\"].replace(\" \",\"_\")\n",
    "    \n",
    "    bodysite_to_run.setdefault(bs,[])\n",
    "    bodysite_to_run[bs].append(row[\"Run\"])\n",
    "    \n",
    "    run_to_bodysite.setdefault(row[\"Run\"],None)\n",
    "    assert run_to_bodysite[row[\"Run\"]] is None or run_to_bodysite[row[\"Run\"]]==bs\n",
    "    run_to_bodysite[row[\"Run\"]]=bs\n",
    "    \n",
    "    tissue_to_bodysite.setdefault(hist_type,[])\n",
    "    tissue_to_bodysite[hist_type].append(bs)\n",
    "    \n",
    "    bodysite_to_tissue.setdefault(bs,None)\n",
    "    assert bodysite_to_tissue[bs] is None or bodysite_to_tissue[bs]==hist_type\n",
    "    bodysite_to_tissue[bs]=hist_type\n",
    "    \n",
    "print(\"number of bodysites: \"+str(len(bodysite_to_run)))\n",
    "\n",
    "for bs in bodysite_to_run:\n",
    "    print(bs,len(bodysite_to_run[bs]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0dcb1e8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get all_id mapping from chess gtf\n",
    "\n",
    "all2chs = dict()\n",
    "\n",
    "with open(chess_gtf_fname,\"r\") as f:\n",
    "    for line in f:\n",
    "        if line.startswith(\"#\"):\n",
    "            continue\n",
    "        lcs = line.strip().split(\"\\t\")\n",
    "        if lcs[0] != \"chr21\":\n",
    "            continue\n",
    "        if lcs[2] == \"transcript\":\n",
    "            aid = None\n",
    "            tid = None\n",
    "            attrs = lcs[8].strip().split(\";\")\n",
    "            for attr in attrs:\n",
    "                attr = attr.strip()\n",
    "                k,v = attr.split(\" \\\"\",1)\n",
    "                v = v.strip('\"')\n",
    "                if k == \"assembly_id\":\n",
    "                    aid = v\n",
    "                if k == \"transcript_id\":\n",
    "                    tid = v\n",
    "            if aid is not None and tid is not None:\n",
    "                all2chs[aid] = tid\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d121fbbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# iterate over the expression file, check if all id is in the all2chs dict\n",
    "# if exists - group expressions by bodysite\n",
    "# if not exists - skip\n",
    "\n",
    "# load all2chs\n",
    "\n",
    "chs_grouped_expressions = dict()\n",
    "\n",
    "with open(non_zero_tpms_fname,\"r\") as f:\n",
    "    for line in f:\n",
    "        lcs = line.strip().split(\"\\t\")\n",
    "        all_id = lcs[0]\n",
    "        if all_id not in all2chs:\n",
    "            continue\n",
    "        \n",
    "        chs_id = all2chs[all_id]\n",
    "        if chs_id not in chs_grouped_expressions:\n",
    "            chs_grouped_expressions[chs_id] = dict()\n",
    "\n",
    "        # iterate over remaining columns - each columns is run:expression value\n",
    "        # run can be looked up in run_to_bodysite\n",
    "\n",
    "        for i in range(1,len(lcs)):\n",
    "            run,expr = lcs[i].split(\":\")\n",
    "            try:\n",
    "                expr = float(expr)\n",
    "            except:\n",
    "                continue\n",
    "            if run not in run_to_bodysite:\n",
    "                continue\n",
    "            bodysite = run_to_bodysite[run]\n",
    "            if bodysite not in chs_grouped_expressions[chs_id]:\n",
    "                chs_grouped_expressions[chs_id][bodysite] = []\n",
    "            chs_grouped_expressions[chs_id][bodysite].append(expr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "2ed25018",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write output file in the following format:\n",
    "# chess_id\\tbodysite1:percentile_data;bodysite2:percentile_data\n",
    "\n",
    "def get_percentiles(data,precision=2):\n",
    "    result = dict()\n",
    "    for p in [25,50,75]:\n",
    "        result[p] = np.round(np.percentile(data, p),precision)\n",
    "    result[\"min\"] = np.round(np.min(data),precision)\n",
    "    result[\"max\"] = np.round(np.max(data),precision)\n",
    "    result[\"mean\"] = np.round(np.mean(data),precision)\n",
    "    result[\"std\"] = np.round(np.std(data),precision)\n",
    "    result[\"count\"] = len(data)\n",
    "    return result\n",
    "\n",
    "with open(\"./chess_expression.chr21.tsv\",\"w\") as outFP:\n",
    "    for chs_id in chs_grouped_expressions:\n",
    "        outFP.write(chs_id+\"\\t\")\n",
    "        for bodysite in chs_grouped_expressions[chs_id]:\n",
    "            percentile_data = get_percentiles(chs_grouped_expressions[chs_id][bodysite])\n",
    "            percentile_data_str = \",\".join([f\"{k}:{v}\" for k,v in percentile_data.items()])\n",
    "            outFP.write(bodysite+\":\"+percentile_data_str+\";\")\n",
    "        outFP.write(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "810c4f88",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "773008be",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "681ff23d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "CHESS_WEB",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
