{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import copy\n",
    "import json\n",
    "import argparse\n",
    "import subprocess\n",
    "\n",
    "import mysql.connector\n",
    "\n",
    "from itertools import chain, combinations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.insert(0, \"../\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "from definitions import *\n",
    "import api"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Connected to MySQL database\n"
     ]
    }
   ],
   "source": [
    "CHESSDB_HOST = \"localhost\"\n",
    "CHESSDB_USER = \"chess_master\"\n",
    "CHESSDB_PASSWORD = \"qwerty\"\n",
    "CHESSDB_PORT = \"3306\"\n",
    "CHESSDB_NAME = \"CHESS_DB\"\n",
    "\n",
    "chessApi = None\n",
    "try:\n",
    "    chessApi = api.CHESS_DB_API(CHESSDB_HOST, CHESSDB_USER, CHESSDB_PASSWORD, CHESSDB_NAME, CHESSDB_PORT)\n",
    "    chessApi.connect()\n",
    "except:\n",
    "    print(\"Failed to connect to database\")\n",
    "    exit(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pseudocode for the SQL query\n",
    "# tid     source            akey   avalue\n",
    "# 1       \"CHESS\"        \"key1\"      \"val4\"\n",
    "# 2       \"CHESS\"        \"key1\"      \"val4\"\n",
    "# 1       \"MANE\"        \"key1\"      \"val2\"\n",
    "# 1       \"CHESS\"        \"key1\"      \"val4\"\n",
    "# 1       \"MANE\"        \"key1\"      \"val2\"\n",
    "# 2       \"CHESS\"        \"key1\"      \"val4\"\n",
    "# 1       \"MANE\"        \"key1\"      \"val2\"\n",
    "# 1       \"CHESS\"        \"key1\"      \"val4\"\n",
    "# 2       \"CHESS\"        \"key1\"      \"val4\"\n",
    "# select first(*) from table order by tid, akey, ORDER BY FIELD(source, 'MANE', 'CHESS', 'RefSeq');\n",
    "#---------------------------------------------------------------------------------------------------\n",
    "\n",
    "from collections import defaultdict\n",
    "\n",
    "# Create the SQL query string\n",
    "# SQL query using a subquery to reference the newAttributes table\n",
    "query = '''SELECT na.*, t.sequenceID, t.start, t.end, t.strand, tx.score, t.exons, tx.cds\n",
    "FROM (\n",
    "    SELECT *\n",
    "    FROM (\n",
    "        SELECT a.*, s.name AS source_name,\n",
    "            ROW_NUMBER() OVER(PARTITION BY a.tid, a.name ORDER BY a.tid, a.name,\n",
    "                CASE\n",
    "                    WHEN s.name = 'CHESS.3.0' THEN 0\n",
    "                    WHEN s.name = 'RefSeq' THEN 1\n",
    "                    WHEN s.name = 'MANE' THEN 2\n",
    "                    ELSE 3  -- Place other sources at the end\n",
    "                END) AS rn\n",
    "        FROM Attributes a\n",
    "        JOIN Sources s ON a.sourceID = s.sourceID\n",
    "    ) AS ranked\n",
    "    WHERE rn = 1 AND (source_name = 'CHESS.3.0' OR source_name = 'RefSeq' OR source_name = 'MANE')\n",
    ") na \n",
    "JOIN TxDBXREF tx ON na.tid = tx.tid AND na.sourceID = tx.sourceID AND na.transcript_id = tx.transcript_id \n",
    "JOIN Transcripts t ON tx.tid = t.tid\n",
    "'''\n",
    "\n",
    "\n",
    "# Define the list of source names for filtering\n",
    "data_sources = ['CHESS.3.0', 'RefSeq', 'MANE']\n",
    "sources_condition = \" OR \".join([f\"source_name = '{source}'\" for source in data_sources])\n",
    "keys = [\"gene_type\"]\n",
    "values = [\"protein_coding\"]\n",
    "\n",
    "where_clause = \"\"\n",
    "if keys and values:\n",
    "    conditions = []\n",
    "    for key, value in zip(keys, values):\n",
    "        conditions.append(f\"(name = '{key}' AND value = '{value}')\")\n",
    "    where_clause = \"WHERE \" + \" OR \".join(conditions)\n",
    "\n",
    "test_query = f'''SELECT na.*, t.sequenceID, t.start, t.end, t.strand, tx.score, t.exons, tx.cds\n",
    "    FROM (\n",
    "        SELECT *\n",
    "        FROM (\n",
    "            SELECT a.*, s.name AS source_name,\n",
    "                ROW_NUMBER() OVER(PARTITION BY a.tid, a.name ORDER BY a.tid, a.name,\n",
    "                    CASE {''.join([f\"WHEN s.name = '{source}' THEN {index} \" for index, source in enumerate(data_sources)])}\n",
    "                    ELSE {len(data_sources)}  -- Place other sources at the end\n",
    "                END) AS rn\n",
    "            FROM Attributes a\n",
    "            JOIN Sources s ON a.sourceID = s.sourceID\n",
    "        ) AS ranked\n",
    "        WHERE rn = 1 AND ({sources_condition})\n",
    "    ) na \n",
    "    JOIN TxDBXREF tx ON na.tid = tx.tid AND na.sourceID = tx.sourceID AND na.transcript_id = tx.transcript_id \n",
    "    JOIN Transcripts t ON tx.tid = t.tid\n",
    "    {where_clause}\n",
    "    '''\n",
    "\n",
    "\n",
    "add_to_the_end = \"\"\"WHERE (name = 'gene_type' AND value = 'protein_coding')\"\"\"\n",
    "\n",
    "# Execute Query\n",
    "select_res = chessApi.execute_query(test_query)\n",
    "select_res\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dictionary mapping transcript id to the row in the SQL query output\n",
    "transcript_data = defaultdict(list) # automatically initializes a list for each key\n",
    "for row in select_res:\n",
    "    transcript_id = row[0]  # Assuming transcript_id is in the first position\n",
    "    transcript_data[transcript_id].append(row)\n",
    "\n",
    "# Writing the GTF/GFF file\n",
    "# Output File Name\n",
    "output_file_type = \"gff\"\n",
    "output_file_name = \"test\"\n",
    "\n",
    "outfname = f\"{output_file_name}.{output_file_type}\"\n",
    "\n",
    "with open(outfname, \"w\") as outFP:\n",
    "    for transcript_id, rows in transcript_data.items():\n",
    "        output_str = \"\"\n",
    "        # Construct transcript line\n",
    "        attribute_list = [(row[3], row[4].replace(\"\\t\", \"\")) for row in rows if row[3] != \"transcript_id\"]\n",
    "\n",
    "        if (output_file_type==\"gtf\"):\n",
    "            attributes = \"; \".join(f\"{attr[0]} \\\"{attr[1]}\\\"\" for attr in attribute_list)\n",
    "\n",
    "        if(output_file_type==\"gff\"):\n",
    "            attributes = \";\".join(f\"{attr[0]}={attr[1]}\" for attr in attribute_list)\n",
    "\n",
    "        # print(attributes + \"\\n\")\n",
    "        strand = \"+\" if rows[0][10] == 1 else \"-\"\n",
    "\n",
    "        # main line for constucting the gtf/gff file\n",
    "        if (output_file_type==\"gtf\"):\n",
    "            output_str = f\"{rows[0][7]}\\t{rows[0][5]}\\ttranscript\\t{rows[0][8]}\\t{rows[0][9]}\\t.\\t{strand}\\t.\\ttranscript_id \\\"{rows[0][2]}\\\"; {attributes}\\n\"\n",
    "\n",
    "        if(output_file_type==\"gff\"):\n",
    "            output_str = f\"{rows[0][7]}\\t{rows[0][5]}\\ttranscript\\t{rows[0][8]}\\t{rows[0][9]}\\t.\\t{strand}\\t.\\ttranscript_id={rows[0][2]};{attributes}\\n\"\n",
    "\n",
    "        outFP.write(output_str)\n",
    "\n",
    "        # construct exon lines\n",
    "        output_str = \"\"\n",
    "\n",
    "        # print exons\n",
    "        if (rows[0][12] != None):\n",
    "            for exon in rows[0][12].split(\",\"):\n",
    "                exon_start, exon_end = [str(int(v)) for v in exon.split(\"-\")]\n",
    "                output_str += rows[0][7]+\"\\t\"+str(rows[0][5])+\"\\texon\\t\"+exon_start+\"\\t\"+exon_end+\"\\t.\\t\"+strand+\"\\t.\\ttranscript_id \\\"\"+str(rows[0][2])+\"\\\";\\n\"\n",
    "            \n",
    "        # print CDS\n",
    "        if (rows[0][13] != None):\n",
    "            for cds in rows[0][13].split(\",\"):\n",
    "                cds_start, cds_end = [str(int(v)) for v in cds.split(\"-\")]\n",
    "                output_str += rows[0][7]+\"\\t\"+str(rows[0][5])+\"\\tCDS\\t\"+cds_start+\"\\t\"+cds_end+\"\\t.\\t\"+strand+\"\\t.\\ttranscript_id \\\"\"+str(rows[0][2])+\"\\\";\\n\"\n",
    "\n",
    "        outFP.write(output_str)\n",
    "\n",
    "\n",
    "        # indexes 0     1       2     3         4           5          6       7     8     9      10    11     12     13\n",
    "        # values tid sourceId tx_id atr_name attr_value source_name sourceID seqId  start  end  strand score  exons  cds\n",
    "\n",
    "# chessApi.disconnect() # Preform when done working with the notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "#####################################################\n",
    "############  EXTRACT GTF WITH INPUTS    ############\n",
    "#####################################################\n",
    "from collections import defaultdict\n",
    "\n",
    "\"\"\"\n",
    "  @brief: Returns a GTF/GFF file from the database using the given inputs\n",
    "\n",
    "  @param: self: CHESS_DB_API object\n",
    "  @param: data_sources: list of data sources to extract from the database (in order of priority)\n",
    "  @param: keys: list of attribute keys/names\n",
    "  @param: values: list of attribute values\n",
    "  @param: nomenclature: nomenclature of the sequenceID\n",
    "  @param: output_file_type: output file type (\"gff\" or \"gtf\")\n",
    "  @param: output_file_name: output file name\n",
    "  @return: GTF/GFF file that meets input criteria\n",
    "\"\"\"\n",
    "def to_gtf_with_inputs(self,data_sources:list,keys:list, values:list, nomenclature:str, output_file_type: str, output_file_name:str):\n",
    "    \n",
    "    ####################\n",
    "    # SQL Query\n",
    "    ####################\n",
    "\n",
    "    # Create SQL clause for data sources\n",
    "    sources_condition = \" OR \".join([f\"source_name = '{source}'\" for source in data_sources])\n",
    "    \n",
    "    # Create SQL clause for attributes\n",
    "    where_clause = \"\"\n",
    "    if keys and values:\n",
    "        conditions = []\n",
    "        for key, value in zip(keys, values):\n",
    "            conditions.append(f\"(name = '{key}' AND value = '{value}')\")\n",
    "        where_clause = \"WHERE \" + \" OR \".join(conditions)\n",
    "\n",
    "    # SQL query\n",
    "    query = f'''SELECT na.*, t.sequenceID, t.start, t.end, t.strand, tx.score, t.exons, tx.cds, t.assemblyName\n",
    "    FROM (\n",
    "        SELECT *\n",
    "        FROM (\n",
    "            SELECT a.*, s.name AS source_name,\n",
    "                ROW_NUMBER() OVER(PARTITION BY a.tid, a.name ORDER BY a.tid, a.name,\n",
    "                    CASE {''.join([f\"WHEN s.name = '{source}' THEN {index} \" for index, source in enumerate(data_sources)])}\n",
    "                    ELSE {len(data_sources)}  -- Place other sources at the end\n",
    "                END) AS rn\n",
    "            FROM Attributes a\n",
    "            JOIN Sources s ON a.sourceID = s.sourceID\n",
    "        ) AS ranked\n",
    "        WHERE rn = 1 AND ({sources_condition})\n",
    "    ) na \n",
    "    JOIN TxDBXREF tx ON na.tid = tx.tid AND na.sourceID = tx.sourceID AND na.transcript_id = tx.transcript_id \n",
    "    JOIN Transcripts t ON tx.tid = t.tid\n",
    "    {where_clause}\n",
    "    '''\n",
    "\n",
    "    # Execute Query\n",
    "    select_res = self.execute_query(query)\n",
    "    \n",
    "\n",
    "    \n",
    "    # Find sequenceId for a given nomenclature\n",
    "    # Execute the SQL query\n",
    "    query = \"SELECT assemblyName, sequenceID, nomenclature, alternativeID FROM SequenceIDMap\"\n",
    "    sid = chessApi.execute_query(query)\n",
    "\n",
    "    # Create a dictionary mapping assemblyName, sequenceId, and nomenclature to alternativeID\n",
    "    sequence_id_map = {}\n",
    "    for row in sid:\n",
    "        key = (row[0], row[1], row[2])  # Use assemblyName, sequenceID, and nomenclature as the key\n",
    "        value = row[3]  # alternativeID as the value\n",
    "        if key not in sequence_id_map:\n",
    "            sequence_id_map[key] = []\n",
    "        sequence_id_map[key].append(value)\n",
    "\n",
    "\n",
    "    ####################\n",
    "    # Create GFF/GTF file\n",
    "    ####################\n",
    "\n",
    "    # dictionary mapping transcript id (tid) to the row in the SQL query output\n",
    "    transcript_data = defaultdict(list) # automatically initializes a list for each key\n",
    "    for row in select_res:\n",
    "        transcript_id = row[0]  # Assuming transcript_id is in the first position\n",
    "        transcript_data[transcript_id].append(row)\n",
    "\n",
    "\n",
    "\n",
    "    # Dictionary to store distinct sources for each transcript Id\n",
    "    distinct_sources = {}\n",
    "\n",
    "    # Iterate through the dictionary\n",
    "    for idx, values in transcript_data.items():\n",
    "        sources_set = set()  # Create a set to collect distinct sources for each index\n",
    "        for data_tuple in values:\n",
    "            source = data_tuple[5]  # Assuming index 5 contains the source information\n",
    "            sources_set.add(source)\n",
    "        distinct_sources[idx] = list(sources_set) \n",
    "\n",
    "\n",
    "    # Writing the GTF/GFF file\n",
    "    outfname = f\"{output_file_name}.{output_file_type}\"\n",
    "\n",
    "    with open(outfname, \"w\") as outFP:\n",
    "        # For each transcript id (actually tid) and all the rows associated with it\n",
    "        for transcript_id, rows in transcript_data.items():\n",
    "            output_str = \"\"\n",
    "            # Construct transcript line\n",
    "            attribute_list = [(row[3], row[4].replace(\"\\t\", \"\")) for row in rows if row[3] not in [\"transcript_id\", \"transcriptId\"]]\n",
    "            attribute_list.append((\"CHESS Database Transcript Sources\", \", \".join(distinct_sources[transcript_id])))\n",
    "\n",
    "            if (output_file_type==\"gtf\"):\n",
    "                attributes = \"; \".join(f\"{attr[0]} \\\"{attr[1]}\\\"\" for attr in attribute_list)\n",
    "\n",
    "            if(output_file_type==\"gff\"):\n",
    "                attributes = \";\".join(f\"{attr[0]}={attr[1]}\" for attr in attribute_list)\n",
    "\n",
    "            # print(attributes + \"\\n\")\n",
    "            strand = \"+\" if rows[0][10] == 1 else \"-\"\n",
    "\n",
    "            # find sequenceId (assemblyId, sequenceId, nomencalture --> alternativeID)\n",
    "            sequence_id = sequence_id_map[(rows[0][14], rows[0][7], nomenclature)][0]\n",
    "\n",
    "            # main line for constucting the gtf/gff file\n",
    "            if (output_file_type==\"gtf\"):\n",
    "                output_str = f\"{sequence_id}\\tCHESS Database\\ttranscript\\t{rows[0][8]}\\t{rows[0][9]}\\t.\\t{strand}\\t.\\ttranscript_id \\\"{rows[0][2]}\\\"; {attributes}\\n\"\n",
    "\n",
    "            if(output_file_type==\"gff\"):\n",
    "                output_str = f\"{sequence_id}\\tCHESS Database\\ttranscript\\t{rows[0][8]}\\t{rows[0][9]}\\t.\\t{strand}\\t.\\ttranscript_id={rows[0][2]};{attributes}\\n\"\n",
    "\n",
    "            outFP.write(output_str)\n",
    "\n",
    "            # construct exon lines\n",
    "            output_str = \"\"\n",
    "\n",
    "            # print exons\n",
    "            if (rows[0][12] != None):\n",
    "                for exon in rows[0][12].split(\",\"):\n",
    "                    exon_start, exon_end = [str(int(v)) for v in exon.split(\"-\")]\n",
    "                    output_str += sequence_id+\"\\t\"+\"CHESS Database\"+\"\\texon\\t\"+exon_start+\"\\t\"+exon_end+\"\\t.\\t\"+strand+\"\\t.\\ttranscript_id \\\"\"+str(rows[0][2])+\"\\\";\\n\"\n",
    "                \n",
    "            # print CDS\n",
    "            if (rows[0][13] != None):\n",
    "                for cds in rows[0][13].split(\",\"):\n",
    "                    cds_start, cds_end = [str(int(v)) for v in cds.split(\"-\")]\n",
    "                    output_str += sequence_id+\"\\t\"+\"CHESS Database\"+\"\\tCDS\\t\"+cds_start+\"\\t\"+cds_end+\"\\t.\\t\"+strand+\"\\t.\\ttranscript_id \\\"\"+str(rows[0][2])+\"\\\";\\n\"\n",
    "\n",
    "            outFP.write(output_str)\n",
    "    return\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#####################################################\n",
    "############  EXTRACT GTF WITH INPUTS    ############\n",
    "#####################################################\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CALLING THE FUNCTION\n",
    "data_sources = ['CHESS.3.0','RefSeq','MANE']\n",
    "keys = []\n",
    "values = []\n",
    "\n",
    "to_gtf_with_inputs(chessApi,data_sources,keys,values,\"GENBANK\",\"gtf\",\"NORMAL\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{('CHM13', 'chr18', 'UCSC'): ['chr18'], ('CHM13', 'chr18', 'GENBANK'): ['CP068260.2'], ('CHM13', 'chr18', 'RefSeq'): ['NC_060942.1'], ('CHM13', 'chr19', 'UCSC'): ['chr19'], ('CHM13', 'chr19', 'GENBANK'): ['CP068259.2'], ('CHM13', 'chr19', 'RefSeq'): ['NC_060943.1'], ('GRCh38', 'chr18', 'UCSC'): ['chr18'], ('GRCh38', 'chr18', 'GENBANK'): ['CM000680.2'], ('GRCh38', 'chr18', 'RefSeq'): ['NC_000018.10'], ('GRCh38', 'chr19', 'UCSC'): ['chr19'], ('GRCh38', 'chr19', 'GENBANK'): ['CM000681.2'], ('GRCh38', 'chr19', 'RefSeq'): ['NC_000019.10']}\n"
     ]
    }
   ],
   "source": [
    "# Add on: sequenceId map\n",
    "\n",
    "# Execute the SQL query\n",
    "query = \"SELECT assemblyName, sequenceID, nomenclature, alternativeID FROM SequenceIDMap\"\n",
    "sid = chessApi.execute_query(query)\n",
    "\n",
    "# Create a dictionary mapping assemblyName, sequenceId, and nomenclature to alternativeID\n",
    "sequence_id_map = {}\n",
    "for row in sid:\n",
    "    key = (row[0], row[1], row[2])  # Use assemblyName, sequenceID, and nomenclature as the key\n",
    "    value = row[3]  # alternativeID as the value\n",
    "    if key not in sequence_id_map:\n",
    "        sequence_id_map[key] = []\n",
    "    sequence_id_map[key].append(value)\n",
    "\n",
    "print(sequence_id_map)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
